### ISSUE TITLE
Fetching from index returns wrong result

### ISSUE THREAD

**Issue by syzdykov** [orientechnologies/orientdb#6727]
## OrientDB Version, operating system, or hardware.
- v2.2 SNAPSHOT

## Operating System
- [x] Linux
- [x] MacOSX

## Expected behavior and actual behavior
I tried to fetch some data from my database and faced with problem.
orientdb {db=mydb}> select count(*) from Movie where releaseDate is null

+----+-----+
|#   |count|
+----+-----+
|0   |128  |
+----+-----+

1 item(s) found. Query executed in 1.104 sec(s).
orientdb {db=mydb}> select count(*) from Movie where releaseDate is null order by releaseDate

+----+-----+
|#   |count|
+----+-----+
|0   |0    |
+----+-----+

It seems "order by" changed result rows for this request. According to "explain" command I found out that second request uses index, so I tried to rebuild this index, but it didn't help me.
orientdb {db=mydb}> REBUILD INDEX Movie_releaseDate

Rebuilding index(es)...
Rebuilt index(es). Found 16332 link(s) in 2.265000 sec(s).

Index(es) rebuilt successfully
orientdb {db=mydb}> select count(*) from Movie where releaseDate is null order by releaseDate

+----+-----+
|#   |count|
+----+-----+
|0   |0    |
+----+-----+

Then I tried to recreate index and this helped me.

orientdb {db=mydb}> DROP INDEX Movie_releaseDate

Removing index...
Dropped index in 0.017000 sec(s).

Index removed successfully
orientdb {db=mydb}> CREATE INDEX Movie_releaseDate On Movie (releaseDate) NOTUNIQUE

Creating index...
Created index successfully with 16332 entries in 1.868000 sec(s).

Index created successfully

orientdb {db=mydb}> select count(*) from Movie where releaseDate is null order by releaseDate

+----+-----+
|#   |count|
+----+-----+
|0   |128  |
+----+-----+

First time this index was created through the createIndex method of OClass from java application.
movie.createIndex("Movie_releaseDate", OClass.INDEX_TYPE.NOTUNIQUE, "releaseDate");

Expected behaviour - fetching from index should return correct result

**Comment by luigidellaquila** [comment]
Hi @syzdykov 

it seems to be a bug. From what I see, probably the first time the index was incorrectly configured to ignore null values and it was (again incorrectly) used to optimize the count().
This index behavior was the default in v 2.1, but it was changed in 2.2.

I'm checking it, I'll let you know asap

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @syzdykov 

I tried to reproduce che problem with a test case, but without success:

```
  @Test
  public void testIndexWithNullValues() {
    OSchemaProxy schema = database.getMetadata().getSchema();

    final OClass cls = schema.createClass("testIndexWithNullValues");
    cls.createProperty("releaseDate", OType.DATE);
    cls.createIndex("testIndexWithNullValues_releaseDate", OClass.INDEX_TYPE.NOTUNIQUE, "releaseDate");


    database.command(new OCommandSQL("insert into testIndexWithNullValues set name = 'foo', releaseDate = null")).execute();
    database.command(new OCommandSQL("insert into testIndexWithNullValues set name = bar")).execute();

    //test with index created from Java
    List<OIdentifiable> result = database.query(
        new OSQLSynchQuery<OIdentifiable>("select count(*) from testIndexWithNullValues where releaseDate is null order by releaseDate"));
    Assert.assertEquals(result.size(), 1);
    Assert.assertEquals(((ODocument)result.get(0)).field("count"), 2L);

    //test with index created from SQL
    result = database.query(
        new OSQLSynchQuery<OIdentifiable>("select count(*) from testIndexWithNullValues where releaseDate is null order by releaseDate"));
    Assert.assertEquals(result.size(), 1);
    Assert.assertEquals(((ODocument)result.get(0)).field("count"), 2L);
  }
```

I'm on 2.2.x branch and everything seems to work fine.
Is it possible that you are using a DB that you created with an old version (eg. a 2.1.x)?

Thanks

Luigi

**Comment by syzdykov** [comment]
Hi @luigidellaquila 
I am not sure about what orientdb version was used for creating this index (it can be 2.1, I am going to ask my colleagues about this). 
But using index which ignores null values to this request looks like a bug

**Comment by syzdykov** [comment]
Yes, it was 2.1 when we create index first time

**Comment by luigidellaquila** [comment]
Ok, thanks for confirming this. So I have only one bug to fix ;-)

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @syzdykov 

I just pushed a fix on 2.2.x branch, it will be released with 2.2.11.
Closing, please reopen if you see any related problems

Thanks

Luigi
